src/ray/core_worker/experimental_mutable_object_provider.cc

```diff
@@ -109,26 +109,44 @@ void MutableObjectProvider::HandlePushMutableObject(
     RAY_CHECK(it != remote_writer_object_to_local_reader_.end());
     info = it->second;
   }
-  size_t data_size = request.data_size();
-  size_t metadata_size = request.metadata_size();
-
-  // Copy both the data and metadata to a local channel.
-  std::shared_ptr<Buffer> data;
-  const uint8_t *metadata_ptr =
-      reinterpret_cast<const uint8_t *>(request.data().data()) + request.data_size();
-  RAY_CHECK_OK(object_manager_->WriteAcquire(info.local_object_id,
-                                             data_size,
-                                             metadata_ptr,
-                                             metadata_size,
-                                             info.num_readers,
-                                             data));
-  RAY_CHECK(data);
-
-  size_t total_size = data_size + metadata_size;
+  size_t total_data_size = request.total_data_size();
+  size_t total_metadata_size = request.total_metadata_size();
+
+  uint64_t written_so_far = request.written_so_far();
+  uint64_t chunk_size = request.chunk_size();
+
+  std::shared_ptr<Buffer> object_backing_store;
```

Comment:
What will happen if the reader doesn't receive the RPC calls in order? For example, if there are 5 chunks, and the reader receives the chunks in the following order: chunk1, chunk5, chunk2, chunk3, and chunk4. Will it call WriteRelease when it receives chunk5?

Reply:
This won't happen. The next chunk is not sent until the RPC response for the current chunk is received.

Reply:
Where is the logic? I think `grpc_client_->PushMutableObject` is an async operation.

Reply:
Oh whoops, you are right. This code should still be correct though because the receiver node should have a single event manager thread that processes the RPCs in the order they arrive.

Reply:
Will the RPCs received by the reader be guaranteed to be in the order of chunk1, chunk2, chunk3, chunk4, and chunk5?

Reply:
Yes, TCP provides this guarantee.

Reply:
Actually, as we discussed in person, gRPCs are not guaranteed to be delivered in order. I updated the PR to support out-of-order delivery of chunks.